homie-controller/src/lib.rs

```diff
@@ -617,3 +616,206 @@ impl From<ParseFloatError> for HandleError {
         HandleError::Warning(format!("Invalid float: {}", e))
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use async_channel::Receiver;
+    use rumqttc::{Packet, Request, Subscribe};
+
+    fn make_test_controller() -> (HomieController, Receiver<Request>) {
+        let (requests_tx, requests_rx) = async_channel::unbounded();
+        let (cancel_tx, _cancel_rx) = async_channel::unbounded();
+        let mqtt_client = AsyncClient::from_senders(requests_tx, cancel_tx);
+        let controller = HomieController {
+            base_topic: "base_topic".to_owned(),
+            mqtt_client,
+            devices: Mutex::new(Arc::new(HashMap::new())),
+        };
+        (controller, requests_rx)
+    }
+
+    fn expect_subscribe(request: Request, expected: Subscribe) {
+        if let Request::Subscribe(actual) = request {
+            assert_eq!(actual, expected);
+        } else {
+            panic!("Expected subscribe but got {:?}", request);
+        }
+    }
+
+    async fn publish(
+        controller: &HomieController,
+        topic: &str,
+        payload: &str,
+    ) -> Result<Option<Event>, PollError> {
+        controller
+            .handle_event(Packet::Publish(Publish::new(
+                topic,
+                QoS::AtLeastOnce,
+                payload,
+            )))
+            .await
+    }
+
+    #[tokio::test]
+    async fn controller_start() -> Result<(), Box<dyn std::error::Error>> {
+        let (controller, requests_rx) = make_test_controller();
+
+        // Start discovering.
+        controller.start().await?;
+        expect_subscribe(
+            requests_rx.recv().await?,
+            Subscribe::new("base_topic/+/$homie", QoS::AtLeastOnce),
+        );
+
+        // Discover a new device.
+        assert_eq!(
+            publish(&controller, "base_topic/device_id/$homie", "4.0").await?,
+            Some(Event::DeviceUpdated {
+                device_id: "device_id".to_owned(),
+                has_required_attributes: false
+            })
+        );
+        expect_subscribe(
+            requests_rx.recv().await?,
+            Subscribe::new("base_topic/device_id/+", QoS::AtLeastOnce),
+        );
+        expect_subscribe(
+            requests_rx.recv().await?,
+            Subscribe::new("base_topic/device_id/$fw/+", QoS::AtLeastOnce),
+        );
+        expect_subscribe(
+            requests_rx.recv().await?,
+            Subscribe::new("base_topic/device_id/$stats/+", QoS::AtLeastOnce),
+        );
```

Comment:
(crazy idea): Could we have something like this that doesn't care about ordering?
```suggestion
        expect_subscriptions(
            requests_rx,
            vec!["base_topic/device_id/+", "base_topic/device_id/$fw/+", "base_topic/device_id/$stats/+"],
        );
```

Reply:
Good idea, I'll do that.